### PR TITLE
[SPARK-41949][CORE][PYTHON] Make stage scheduling support local-cluster mode

### PR DESCRIPTION
**Pull request by WeichenXu123**
Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

### What changes were proposed in this pull request?

Make stage scheduling support local-cluster mode.

### Why are the changes needed?

This is useful in testing, especially for test code of third-party python libraries that depends on pyspark, many tests are written with pytest, but pytest is hard to integrate with a standalone spark cluster.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit tests

### CODE REVIEW

**Review by Ngone51 — APPROVED** [review]
Since we have supported resource profile in Standalone with dynamic allocation disabled after SPARK-39853 , it makes sense to me to support "local-cluster" mode too.

**Comment by Ngone51** [comment]
cc @tgravescs @ivoson

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/tests/test_stage_sched.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            cpus_per_worker=4,
+            gpus_per_worker=4,
+            num_tasks=4,
+            resource_profile=rp,
+            expected_max_concurrent_tasks=2,
+        )
+
+
+if __name__ == "__main__":
+    from pyspark.tests.test_memory_profiler import *  # noqa: F401
```
Comment (HyukjinKwon) [inline python/pyspark/tests/test_stage_sched.py:132]:
```suggestion
    from pyspark.tests.test_stage_sched import *  # noqa: F401
```

File: python/pyspark/tests/test_stage_sched.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            os.remove(pid_file_path)
+            return num_concurrent_tasks
+
+        results = (
+            self.sc.parallelize(range(num_tasks), num_tasks)
+            .withResources(resource_profile)
+            .map(mapper)
+            .collect()
+        )
+        assert max(results) == expected_max_concurrent_tasks
```
Comment (HyukjinKwon) [inline python/pyspark/tests/test_stage_sched.py:94]: `self.assertEqual`
```suggestion
        self.assertEqual(max(results), expected_max_concurrent_tasks)
```

File: python/pyspark/tests/test_stage_sched.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            conf.set("spark.worker.resource.gpu.amount", str(gpus_per_worker))
+            conf.set("spark.worker.resourcesFile", worker_res_config_file)
+            conf.set("spark.executor.resource.gpu.amount", str(gpus_per_worker))
+
+        self.sc = SparkContext(conf=conf)
+        pids_output_dir = os.path.join(self.temp_dir, "pids")
+        os.mkdir(pids_output_dir)
+
+        def mapper(_):
+            from pyspark.taskcontext import TaskContext
```
Comment (HyukjinKwon) [inline python/pyspark/tests/test_stage_sched.py:76]: you can just import on the top

File: python/pyspark/tests/test_stage_sched.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/tests/test_stage_sched.py:132]: I would actually name it something like `test_resource`

**Comment by HyukjinKwon** [comment]
some nits. cc @tgravescs FYI.

LGTM

**Review by ivoson — APPROVED** [review]
LGTM

**Pull request closed without merging**